### PR TITLE
Deprecate non-bool masks in masked_fill

### LINKED ISSUE #85063: Tensor.masked_fill treats DoubleTensor masks as BoolTensor masks of all False
### 🐛 Describe the bug

Tensor.masked_fill treats a DoubleTensor mask as a BoolTensor of all False, which as @ramiro050 pointed out [is inconsistent with masked_select](https://github.com/llvm/torch-mlir/pull/1337#discussion_r971193308). Is this expected behavior or should passing a DoubleTensor produce a runtime error like it does for masked_select?

### Versions

PyTorch version: 1.13.0.dev20220913+cpu
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: Debian GNU/Linux 11 (bullseye) (x86_64)
GCC version: (Debian 10.2.1-6) 10.2.1 20210110
Clang version: Could not collect
CMake version: version 3.24.1
Libc version: glibc-2.34

Python version: 3.10.7+ (heads/3.10:4e1303b, Sep 12 2022, 19:13:46) [GCC 10.2.1 20210110] (64-bit runtime)
Python platform: Linux-5.10.0-16-amd64-x86_64-with-glibc2.34
Is CUDA available: False
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] numpy==1.23.3
[pip3] torch==1.13.0.dev20220913+cpu
[pip3] torchvision==0.14.0.dev20220913+cpu
[conda] Could not collect

### PR DESCRIPTION
**Pull request by kvathupo**
__What?__
Per discussion at #94634, deprecate `masked_fill` with non-bool masks. Deprecation warnings were previously added by #22261, but not for Apple MPS. I can revert the MPS changes if deprecation warnings are wanted first tho. See also #96112. 

Fixes #85063 and #89320.

__Further Development?__
- Fixed the mask dtype checking for the cuda dispatch for `masked_fill` in `aten/src/ATen/native/cuda/Indexing.cu`

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @EikanWang @soumith @voznesenskym @penguinwu @anijain2305 @Guobing-Chen @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

### CODE REVIEW

**Comment by kvathupo** [comment]
CC: @malfet @ngimel

File: aten/src/ATen/native/cpu/IndexKernel.cpp [hunk 1]
```diff
@@ -304,15 +304,12 @@ void index_copy_kernel(
 
 template <typename scalar_t, typename mask_t>
```
Comment (malfet) [inline aten/src/ATen/native/cpu/IndexKernel.cpp:305]: Nit, no need to pass `mask_t` as it is fixed
```suggestion
template <typename scalar_t>
```

File: aten/src/ATen/native/cpu/IndexKernel.cpp [hunk 2]
```diff
@@ -304,15 +304,12 @@ void index_copy_kernel(
...
 void cpu_masked_fill_kernel(TensorIterator& iter, scalar_t value) {
-  auto is_mask_bool = std::is_same<mask_t, bool>::value;
   auto loop = [&](char** data, const int64_t* strides, int64_t n) {
     char* dst = data[0];
     char* mask = data[1];
     for (const auto i : c10::irange(n)) {
       mask_t mask_value = *(mask_t*)(mask + strides[1] * i);
```
Comment (malfet) [inline aten/src/ATen/native/cpu/IndexKernel.cpp:311]: Nit (as mask type is always bool
```suggestion
      auto mask_value = *static_cast<bool*>(mask + strides[1] * i);
```

File: aten/src/ATen/native/cpu/IndexKernel.cpp [hunk 3]
```diff
@@ -326,11 +323,9 @@ void masked_fill_kernel(TensorIterator& iter, const Scalar& value) {
...
       scalar_t scalar_val = value.to<scalar_t>();
       auto mask_dtype = iter.input_dtype(0);
-      if (mask_dtype == ScalarType::Bool) {
-        cpu_masked_fill_kernel<scalar_t, bool>(iter, scalar_val);
-      } else {
-        cpu_masked_fill_kernel<scalar_t, unsigned char>(iter, scalar_val);
-      }
+      TORCH_CHECK(mask_dtype == ScalarType::Bool, "masked_fill only supports boolean masks, "
+        "but got mask with dtype ", mask_dtype);
+      cpu_masked_fill_kernel<scalar_t, bool>(iter, scalar_val);
```
Comment (malfet) [inline aten/src/ATen/native/cpu/IndexKernel.cpp:328]:
```suggestion
      cpu_masked_fill_kernel<scalar_t>(iter, scalar_val);
```

File: aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp [hunk 4]
```diff
@@ -4098,15 +4098,12 @@ void dequantize_tensor_per_tensor_affine_sub_byte_cpu(
 // This function expects quantized_val input to already be quantized
 template <typename scalar_t, typename mask_t>
```
Comment (malfet) [inline aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp:4099]: Nit
```suggestion
template <typename scalar_t>
```

File: aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp [hunk 5]
```diff
@@ -4098,15 +4098,12 @@ void dequantize_tensor_per_tensor_affine_sub_byte_cpu(
...
 void cpu_masked_fill_kernel_quantized_cpu(TensorIterator& iter, scalar_t quantized_val) {
-  auto is_mask_bool = std::is_same<mask_t, bool>::value;
   auto loop = [&](char** data, const int64_t* strides, int64_t n) {
     char* dst = data[0];
     char* mask = data[1];
     for (const auto i : c10::irange(n)) {
       mask_t mask_value = *(mask_t*)(mask + strides[1] * i);
```
Comment (malfet) [inline aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp:4105]:
```suggestion
      auto mask_value = *static_cast<mask_t*>(mask + strides[1] * i);
```

File: aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp [hunk 6]
```diff
@@ -4120,11 +4117,9 @@ void masked_fill_kernel_quantized_cpu(TensorIterator& iter, const Scalar& value,
...
     auto quantized_val = quantize_val<scalar_t>(scale, zero_point, float_val);
     auto mask_dtype = iter.input_dtype(0);
-    if (mask_dtype == ScalarType::Bool) {
-      cpu_masked_fill_kernel_quantized_cpu<scalar_t, bool>(iter, quantized_val);
-    } else {
-      cpu_masked_fill_kernel_quantized_cpu<scalar_t, unsigned char>(iter, quantized_val);
-    }
+    TORCH_CHECK(mask_dtype == ScalarType::Bool, "masked_fill only supports boolean masks, "
+      "but got mask with dtype ", mask_dtype);
+    cpu_masked_fill_kernel_quantized_cpu<scalar_t, bool>(iter, quantized_val);
```
Comment (malfet) [inline aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp:4122]:
```suggestion
    cpu_masked_fill_kernel_quantized_cpu<scalar_t>(iter, quantized_val);
```

**Review by malfet — APPROVED** [review]
Thank you very much for the PR. Looks good to me.
Please cleanup `mask_t` template parameter (left a few suggestions, tough technically its optional)

**Comment by kvathupo** [comment]
> Thank you very much for the PR. Looks good to me. Please cleanup `mask_t` template parameter (left a few suggestions, tough technically its optional)

Applied, looks cleaner without the redundant bool params and checks !

**Review by ngimel — APPROVED** [review]
Let's hope nothing breaks

**Pull request closed without merging**